### ISSUE TITLE
Conditional HTTP requests broken?

### ISSUE THREAD

**Issue by ehamberg** [Ranchero-Software/NetNewsWire#4345]
Not any specifics, unfortunately, but from http://rachelbythebay.com/w/2024/08/02/fs/ it sounds like NetNewsWire (or its HTTP library, I guess) is not doing conditional HTTP requests correctly:

> NetNewsWire. Bigtime brokenness in their IMS/INM caching. No idea why, but it makes it generate unconditionals constantly.

> NetNewsWire. Just like the earlier one, broken caching and crazy timing.

> NetNewsWire. Same bad caching and timing stuff.

> NetNewsWire. Broken, as above.

> NetNewsWire. Bro-ken.

> NetNewsWire. Broken record.

**Comment by brentsimmons** [comment]
We already have a bug about OpenRSS and rate limiting: https://github.com/Ranchero-Software/NetNewsWire/issues/4224

And I’m not too concerned about the timing issues in Rachel’s report — you can do a manual refresh any time you want in NetNewsWire, and that will result in a new request. So any crazy timing is just that.

But the conditional GET bugs are concerning. Here’s the actual report: https://rachelbythebay.com/api/fsr/4467c09c31370b8375c6b2a320aa77cd5d999cd4

(I had the test feed running on my personal laptop, and it reflects my use. You can even see when I went on a long vacation and left the laptop at home.)

I’ve identified one issue — I don’t know if it’s the only one or not. Here’s the scoop:

If the server returns a raw feed that is unchanged from the previous request, we ignore any updated Last-Modified or Etag headers. We make the assumption that those headers would not have changed if the raw feed hasn’t changed. This is surely true almost all the time in the real world!

There may be other issues too, of course, and it’s possible my theory isn’t the cause of the issues here. But it’s still a thing to fix.

I’ve asked Rachel to add a column for http response code — I have a suspicion that this might tell me more. Fingers crossed that the data is available and added to the report.

**Comment by brentsimmons** [comment]
Follow-up: I asked Rachel about the http response codes for the test feed — it’s always 200.

Blogged about this: https://inessential.com/2024/08/03/netnewswire_and_conditional_get_issues.html

**Comment by DamonHD** [comment]
You should definitely accept and store new Cache-Control and Expires (and ETag and Retry-After) headers whether or not the body has changed, else you will not correctly put off your next fetch when (say) those Cache-Control / Expires headers are being updated to tell you when it is next worth polling.

My feed dynamically computes those, eg to steer you away from polling when I am asleep and it is dark and there will likely be no updates from my off-grid server and me.

(You may already have fixed this, I hope so.)

**Comment by brentsimmons** [comment]
We store ETag and LastModified for all feeds to do Conditional GET requests.

We store Cache-Control _only_ for OpenRSS feeds. We did it for all feeds for a while, and it caused huge problems because Cache-Control settings for most feeds are completely unintentional, and it sometimes led to days between updates.

**Comment by brentsimmons** [comment]
We also do now store ETag and LastModified even when the body hasn’t changed, yes.

**Comment by DamonHD** [comment]
Hurrah!  The climate and creators' bandwidth bills should thank you!  B^>

Are you now also capturing Cache-Control max-age and Retry-After and treating them as 'do not retry before?

**Comment by DamonHD** [comment]
What bad things are you seeing in Cache-Control?

If you ignore my server's Cache-Control I have no way of telling you when it is sensible for you to poll again.  I really really don't want you polling at night in winter on my off-grid server when I am asleep and will not be generating new episodes.

Does Expires work better?

**Comment by brentsimmons** [comment]
The bad things we’re seeing in Cache-Control are things like refreshing a given feed once every 36 hours, even though it’s a feed with multiple daily updates and everyone, including the feed author, expects RSS readers to fetch it with much greater frequency.

All these RSS feed providers aren’t going to suddenly deal with Cache-Control configuration on their servers. So we can’t use it in general. (We tried! It was terrible.)

However: if you’d like me to add your host to the list where we do pay attention to Cache-Control, just let me know what the host name is. (Right now that list is just openrss.org, but I’m happy to add to it.)

Note, though, that this change wouldn’t take effect until the next releases, and I don’t have specific dates for those releases (or even good guesses).

**Comment by DamonHD** [comment]
My feed has no listeners, so don't bother with anything for my server!

I'm more interested in general solutions that cut the 99%+ bandwidth waste in feeds, which I talked about at LOCO2024 (https://www.sicsa.ac.uk/loco/loco2024/) and FOSDEM this year, and I am right now working on an arXiv / NLUUG paper about!

is NNW capabale of testing if the actual update frequency is much different to Cache-Control / Expires and ONLY in that case, ignoring the bad headers, so that the rest of us can have nice things?

Or a simple exponential back off, eg with a cap of 1h on the retry interval implied by the headers, doubling each time the feed hasn't changed, and resetting to 1h when there is a content change?

**Comment by brentsimmons** [comment]
The first best step, the biggest bandwidth savings that we could get, is to get feed providers to actually handle conditional GET on the server side.

They either don’t provide etag/last-modified or they do, and we pass them back properly — and they just ignore it and respond with a 200 anyway and the body of the feed.

This is in their own best interest! And in ours, too — we obviously want our apps to use the least networking and energy as possible.

But that’s the actual state of things — we’re nowhere near being able to have nicer things than that. (Well, again, I’m happy to create allow-lists for nicer things.)

**Comment by DamonHD** [comment]
Yes, working conditiional GET response would be excellent, but avoiding some pointless requests at all would also help you?

Thank you for this very fast interaction, BTW!

Would the exponential backoff scheme above be do-able?  As long as you are doing conditional requests no closer than 1h, Rachey (by the Bay) is going to be happy too...  B^>

It seems that most Apache servers do not handle ETag correctly (I can give you a link if you wish), so maybe ignore ETag from Apache servers that also offer Last-Modified and see if that is useful?

**Comment by brentsimmons** [comment]
The exponential back-off plan would result in unhappy users who aren’t getting timely updates.

I plan to add a special case for the Rachel by the Bay feed(s). As already done for openrss.org. Again, happy to do this on request for anybody who wants it.

But we are a long, long way from general solutions. Near-universal support of conditional GET would be a huge win.

**Comment by DamonHD** [comment]
Thanks for continuing to hunour me!

Do those feeds that are doing frequent updates AND bad cache-control / expires / last-modified / etag handling contain any of the following fields in the feed itself which you could use as a cue to do different/better things:

ttl (base RSS)
sy:updatePeriod (xmlns:sy="http://purl.org/rss/1.0/modules/syndication/")
sy:updateFrequency
podcast:updateFrequency (xmlns:podcast="https://podcastindex.org/namespace/1.0")

(I'd also like skipHours and and skipDays paid attention to, and I know that I'm not *quite* the only one using them, but...)

I will make more fuss about the conditional GET stuff when talking to feed providers. It would also be great if some (eg Apple) would also allow-encoding of at least gzip given how compressable these feeds are!

**Comment by brentsimmons** [comment]
Those things, including skipHours and skipDays, are included pretty much never. 

NetNewsWire accepts gzip compression.

**Comment by DamonHD** [comment]
Thank you for trying to Do The Right Thing (TM)!

It would be interesting to note if the Apache / ETag suggestion above helps with conditional GETs.  I was staggered that my fairly-standard config had been basically wrong from the start, and I probably put up my first Apache server last century!

https://www.earth.org.uk/note-on-site-technicals-83.html#2024-04-21

Googlebot does lots more 304s now...

**Comment by DamonHD** [comment]
Note: Rachel has an updated document that is worth reading!

https://rachelbythebay.com/frb/
